### ISSUE TITLE
`Writer.append_data` hangs when starting ffmpeg process.

### ISSUE THREAD

**Issue by Bulkin** [imageio/imageio-ffmpeg#58]
Probably related to #33.

Something very weird is happening when `Popen` is called with a `preexec_fn`  in a complex program with many moving parts. The calling thread hangs trying to read from a pipe (presumbaly created by `Popen`, although setting `stdout` and `stderr` to None or DEVNULL has no effect on the behavior), while the new process blocks on a futex:

```
[pid 30794] set_robust_list(0x7f54ba3629e0, 24 <unfinished ...>
[pid 30794] <... set_robust_list resumed> ) = 0
[pid 30794] write(12, "\1\0\0\0\0\0\0\0", 8) = 8
[pid 30794] write(12, "\1\0\0\0\0\0\0\0", 8) = 8
[pid 30794] write(12, "\1\0\0\0\0\0\0\0", 8) = 8
[pid 30794] write(12, "\1\0\0\0\0\0\0\0", 8 <unfinished ...>
[pid 30794] <... write resumed> )       = 8
[pid 30794] write(12, "\1\0\0\0\0\0\0\0", 8 <unfinished ...>
[pid 30794] <... write resumed> )       = 8
[pid 30794] write(12, "\1\0\0\0\0\0\0\0", 8 <unfinished ...>
[pid 30794] <... write resumed> )       = 8
[pid 30794] write(12, "\1\0\0\0\0\0\0\0", 8) = 8
[pid 30794] poll([{fd=10, events=POLLIN|POLLOUT}], 1, -1) = 1 ([{fd=10, revents=POLLOUT}])
[pid 30794] writev(10, [{iov_base="\202\2\2\0\10\0\200\10+\0\1\0", iov_len=12}], 1) = 12
[pid 30794] futex(0x7f54ba35c338, FUTEX_WAIT_PRIVATE, 0, NULL <unfinished ...>
```

This isn't caused by `imageio` specifically, since the hang is reproducible with
```python
def nop(): pass

Popen(["uname", '-a'], stdin=None, stdout=None, stderr=None, preexec_fn=nop)
```

but imageio seems to be the best place for a workaround.

Would you consider merging a pull request adding a `prevent_sigint` argument to `write_frames` (and `read_frames` for symmetry)? It would require changes to `imageio` too, since the arguments are passed explicitly in `FfmpegFormat._initialize`.

**Comment by almarklein** [comment]
Thanks for reporting this!

Just to make sure I understand: normal usage of imageio_ffmpeg is fine, but in a specific application there seems to be some interaction causing threads to stall, and this situation is triggered by imageio_ffmpeg setting `preexec_fn`.

Could you also provide some details on your system?

I'm secretly hoping that there is a way to fix this without adding to the public API. Another solution could be an environment variable. Ideas?

**Comment by Bulkin** [comment]
Your understanding is correct. In fact, this piece of code reproduces the issue:

```python
#!/usr/bin/env python3

from threading import Thread
from PySide2.QtWidgets import QApplication
from PySide2.QtQml import QQmlApplicationEngine

def main():
    def test_popen():
        def nop(): pass
        from subprocess import Popen
        print('TEST START')
        Popen(["uname", "-a"], preexec_fn=None) # This works
        Popen(["uname", "-a"], preexec_fn=nop)  # This hangs
        print('TEST END')

    Thread(target=test_popen).start()

    app = QApplication()
    qmlEngine = QQmlApplicationEngine()

    app.exec_()


if __name__ == '__main__':
    # This is needed for Ctrl-C to work, has no effect on the bug
    import signal
    signal.signal(signal.SIGINT, signal.SIG_DFL)

    main() # Inlined code from main runs without issue
```

The question about the system is an interesting one.  So far, I've managed to reproduce the issue on a single system (where the bug originally appeared), it is Debian buster running on a first gen ryzen with nvidia graphics.

I guess the main question is, can a configurable `prevent_sigint` be beneficial in other usecases? Am I correct in assuming it was initially added, so that ffmpeg can finalize the written file after the calling script was stopped by `Ctrl-C` in a terminal?

**Comment by almarklein** [comment]
> Am I correct in assuming it was initially added, so that ffmpeg can finalize the written file after the calling script was stopped by Ctrl-C in a terminal?

Yes, and in interactive use, you can interrupt the process and then even append more frames. See #4 for details. Basically, we want to prevent interrups to propage to the ffmpeg process because it was most likely meant to interrupt running code in Python.

The reason why its a param in `_popen_kwargs()` is because it's not needed for process calls that return immediately.

My inclination is to define an environment variable to control this behavior, so its easy to turn it off, without polluting the API.

**Comment by Bulkin** [comment]
I was going to say that an env var is at the wrong level for configuring something like that, but having thought about it some more and after some testing, now agree that it is the best solution for the circumstances.
